{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook to analyze and display scRNAseq data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load important modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Standard modules\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd \n",
    "import plotly.graph_objects as go\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import linalg\n",
    "from plotly.subplots import make_subplots\n",
    "import ipywidgets as widgets\n",
    "import logging\n",
    "from sklearn import linear_model\n",
    "from multiprocessing import Pool\n",
    "import plotly.express as px\n",
    "import scipy.cluster.hierarchy as sch\n",
    "from numba import njit\n",
    "from umap import UMAP\n",
    "from sklearn.cluster import KMeans\n",
    "import copy\n",
    "from scipy.spatial.distance import pdist\n",
    "from scipy.cluster.hierarchy import ward, fcluster, leaves_list, optimal_leaf_ordering\n",
    "import subprocess\n",
    "\n",
    "# Move to root directory for easier module handling\n",
    "os.chdir(\"../..\")\n",
    "print(os.listdir(\".\"))\n",
    "\n",
    "#LBAE imports\n",
    "from modules.maldi_data import MaldiData\n",
    "from modules.figures import Figures\n",
    "from modules.atlas import Atlas\n",
    "from modules.launch import Launch\n",
    "from modules.storage import Storage\n",
    "\n",
    "\n",
    "# multithreading/multiprocessing\n",
    "from multiprocessing import Pool\n",
    "import multiprocessing\n",
    "from threadpoolctl import threadpool_limits\n",
    "\n",
    "# set thread limit\n",
    "threadpool_limits(16)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load LBAE objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_data = \"data/whole_dataset/\"\n",
    "path_annotations = \"data/annotations/\"\n",
    "path_db = \"data/app_data/data.db\"\n",
    "\n",
    "# Load shelve database\n",
    "storage = Storage(path_db)\n",
    "\n",
    "# Load data\n",
    "data = MaldiData(path_data, path_annotations)\n",
    "\n",
    "# Load Atlas and Figures objects. At first launch, many objects will be precomputed and shelved in\n",
    "# the classes Atlas and Figures.\n",
    "atlas = Atlas(data, storage, resolution=25, sample = True)\n",
    "figures = Figures(data, storage, atlas, sample = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load genes and lipids arrays "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_exp_lipids = np.load('data/scRNAseq/array_exp_lipids.npy')\n",
    "array_exp_genes = np.load('data/scRNAseq/array_exp_genes.npy')\n",
    "l_name_lipids = np.load('data/scRNAseq/array_name_lipids.npy').tolist()\n",
    "l_genes = np.load('data/scRNAseq/array_name_genes.npy').tolist()\n",
    "xmol, ymol, zmol = np.load('data/scRNAseq/array_coordinates.npy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(array_exp_genes.shape, array_exp_lipids.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make an elatic net regression to explain lipid expression in terms of combination gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define regression as a function for parallelization\n",
    "def compute_regression(index_lipid):\n",
    "    print(index_lipid)\n",
    "    clf = linear_model.ElasticNet(fit_intercept = True, alpha=0.2, positive = False)#linear_model.Lasso(alpha=0.001)##\n",
    "    clf.fit(array_exp_genes, array_exp_lipids[:,index_lipid])\n",
    "    return [clf.coef_, clf.score(array_exp_genes, array_exp_lipids[:,index_lipid])]\n",
    "\n",
    "# Parallelize\n",
    "#with Pool(processes=16) as pool:\n",
    "l_lipid_indices = list(range(array_exp_lipids.shape[1]))\n",
    "l_res =  [x for x in map(compute_regression, l_lipid_indices[:])]\n",
    "\n",
    "ll_coef = []\n",
    "l_score = []\n",
    "for res in l_res:\n",
    "    ll_coef.append(res[0])\n",
    "    l_score.append(res[1])\n",
    "\n",
    "print(l_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(l_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_coef = np.array(ll_coef)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(np.sum([1 for x in l_res[0][0] if abs(x)>0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove genes that do not explain the expression of any lipids\n",
    "l_to_keep = []\n",
    "for idx, col in enumerate(array_coef.T):\n",
    "    if np.sum([1 for x in col if abs(x)>0])>15:\n",
    "        l_to_keep.append(idx)\n",
    "        \n",
    "array_coef = array_coef[:, l_to_keep]\n",
    "array_genes = np.array(l_genes)[l_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(array_coef.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute correlation on rows\n",
    "pairwise_distances = sch.distance.pdist(array_coef)\n",
    "linkage = sch.linkage(pairwise_distances, method='ward')\n",
    "cluster_distance_threshold = pairwise_distances.max()/2\n",
    "idx_to_cluster_array = sch.fcluster(linkage, cluster_distance_threshold, \n",
    "                                    criterion='distance')\n",
    "idx = np.argsort(idx_to_cluster_array)\n",
    "    \n",
    "corr_array_clustered = array_coef[idx, :]\n",
    "\n",
    "# Compute correlation on columns\n",
    "pairwise_distances = sch.distance.pdist(corr_array_clustered.T)\n",
    "linkage = sch.linkage(pairwise_distances, method='ward')\n",
    "cluster_distance_threshold = pairwise_distances.max()/2\n",
    "idx_to_cluster_array = sch.fcluster(linkage, cluster_distance_threshold, \n",
    "                                    criterion='distance')\n",
    "idx = np.argsort(idx_to_cluster_array)\n",
    "    \n",
    "corr_array_clustered = corr_array_clustered[:, idx]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(corr_array_clustered.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = make_subplots(rows=1, cols=2, shared_yaxes=True,horizontal_spacing=0.)\n",
    "\n",
    "\n",
    "g1 = go.Heatmap(z = corr_array_clustered, colorscale='RdBu', x=array_genes, y = l_name_lipids, zmin = -0.002, zmax = 0.002, colorbar=dict(title=\"Coef value\", x=1.1 ))\n",
    "g2 = go.Heatmap(z = np.array([l_score]).T, colorscale='mint', y = l_name_lipids, zmin = 0., zmax = 0.6, colorbar=dict(title=\"R2 score\", x=1.05 ))\n",
    "\n",
    "\n",
    "fig.append_trace(g1, row=1, col=1)\n",
    "fig.append_trace(g2, row=1, col=2)\n",
    "\n",
    "\n",
    "\n",
    "# edit axis labels\n",
    "fig['layout']['xaxis']['title']=\"Genes\"\n",
    "fig['layout']['yaxis']['title']='Lipids'\n",
    "#fig['layout']['xaxis2']['title']='R2 score'\n",
    "fig['layout']['xaxis2']['side']='top'\n",
    "\n",
    "fig.update_xaxes(showticklabels=False, row=1, col=2)\n",
    "fig.update_yaxes(showticklabels=False, row=1, col=2)\n",
    "\n",
    "fig['layout']['yaxis2']['scaleanchor']='x2'\n",
    "fig['layout']['yaxis']['scaleanchor']='x'\n",
    "\n",
    "fig['layout']['xaxis']['domain'] = [0, 0.98]\n",
    "fig['layout']['xaxis2']['domain'] = [0.99, 1]\n",
    "#fig.update_traces(showscale=False)\n",
    "fig.update_xaxes(tickangle=45)\n",
    "#fig.update_layout(coloraxis_colorbar_x=-0.15)\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    width=int(1520),\n",
    "    height=int(700),\n",
    "    font_size=7,\n",
    "    title_font_size=12,\n",
    "    title={\n",
    "        'text' : 'Elastic Net regression factors (MALDI lipids data with gene expression as predictors)',\n",
    "        'y':0.92,\n",
    "        'x':0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top'}\n",
    ")\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "\n",
    "\n",
    "fig.show()\n",
    "fig.write_image(\"notebooks/scRNAseq/heatmap.pdf\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make an interactive widget displaying a clustergram of lipid and gene expression when one clicks on the corresponding dot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get root figure\n",
    "root_data = figures._storage.return_shelved_object(\n",
    "    \"figures/3D_page\",\n",
    "    \"volume_root\",\n",
    "    force_update=False,\n",
    "    compute_function=figures.compute_3D_root_volume,\n",
    ")\n",
    "\n",
    "# Get scatter figure for the scRNAseq spots\n",
    "scatter = go.Scatter3d(\n",
    "    x=xmol,\n",
    "    y=ymol,\n",
    "    z=zmol,\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        size=2,\n",
    "        opacity=0.8\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "root_data['hoverinfo']= 'skip'\n",
    "scatter['hoverinfo']= \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_widget = True\n",
    "f = None\n",
    "if plot_widget:\n",
    "    fig = make_subplots(rows=1, cols=6, specs=[[{'type': 'surface', \"colspan\": 4},  {'type': 'xy'}, {'type': 'xy'}, {'type': 'xy'}, {'type': 'xy'}, {'type': 'xy'}]])\n",
    "    fig.add_trace(root_data,row=1, col=1)\n",
    "    fig.add_trace(scatter,row=1, col=1)\n",
    "\n",
    "    n_to_keep = 53\n",
    "\n",
    "    # Sort array according to most expressed genes and lipid for spot 0\n",
    "    index_init = 0\n",
    "    array_init_genes, l_genes_init = zip(*sorted(zip(array_exp_genes[index_init:index_init+1,:].T, l_genes), reverse=True))\n",
    "    array_init_genes = np.array(array_init_genes)[:n_to_keep]\n",
    "    l_genes_init = l_genes_init[:n_to_keep]\n",
    "    array_init_lipids, l_lipids_init = zip(*sorted(zip(array_exp_lipids[index_init:index_init+1,:].T, l_name_lipids), reverse=True))\n",
    "    array_init_lipids = np.array(array_init_lipids)[:n_to_keep]\n",
    "    l_lipids_init = l_lipids_init[:n_to_keep]\n",
    "\n",
    "    g1 = go.Heatmap(z = array_init_genes, colorscale='mint', y=l_genes_init)#, xaxis=\"x2\", yaxis=\"y2\", )\n",
    "    g2 = go.Heatmap(z = array_init_lipids, colorscale='mint', y = l_lipids_init )#, xaxis=\"x3\", yaxis=\"y3\", )\n",
    "\n",
    "\n",
    "    fig.add_trace(g1, row=1, col=5)\n",
    "    fig.add_trace(g2, row=1, col=6)\n",
    "\n",
    "\n",
    "    fig.data[2].showscale=False\n",
    "    fig.data[3].showscale=False\n",
    "\n",
    "\n",
    "\n",
    "    #fig.add_trace(go.Scatter(x=[20, 30, 40], y=[50, 60, 70]),row=1, col=2)\n",
    "\n",
    "    f = go.FigureWidget(fig)\n",
    "    p1 = f.data[1]\n",
    "    p2 = f.data[2]\n",
    "    p3 = f.data[3]\n",
    "\n",
    "\n",
    "\n",
    "    # Hide grey background\n",
    "    f.update_layout(\n",
    "        margin=dict(t=0, r=0, b=0, l=0),\n",
    "        scene=dict(\n",
    "            xaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\"),\n",
    "            yaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\"),\n",
    "            zaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\"),\n",
    "        ),\n",
    "        template = \"plotly_dark\",\n",
    "        plot_bgcolor = \"rgba(0,0,0,0)\",\n",
    "        paper_bgcolor = \"rgba(0,0,0,0)\",\n",
    "\n",
    "        xaxis=dict(\n",
    "                domain=[0.7, 0.75]\n",
    "            ),\n",
    "        xaxis2=dict(\n",
    "                domain=[0.90, 0.95]\n",
    "            ),\n",
    "        #xaxis3=dict(\n",
    "        #        domain=[0.75, 1],\n",
    "        #),\n",
    "    )\n",
    "\n",
    "    f['layout']['yaxis2']['scaleanchor']='x2'\n",
    "    f['layout']['yaxis']['scaleanchor']='x'\n",
    "\n",
    "\n",
    "\n",
    "    f.update_layout(\n",
    "        #width=int(1200*1.2),\n",
    "        #height=int(700*1.2),\n",
    "        font_size=7,\n",
    "        title_font_size=12,\n",
    "        title={\n",
    "            'text' : 'Click on a dot to display the corresponding data',\n",
    "            'y':0.92,\n",
    "            'x':0.3,\n",
    "            'xanchor': 'center',\n",
    "            'yanchor': 'top'}\n",
    "    )\n",
    "\n",
    "\n",
    "    #print(f['layout'])\n",
    "    #f['layout']['xaxis']['domain'] = [0, 0.96]\n",
    "    #f['layout']['xaxis2']['domain'] = [0.96, 0.98]\n",
    "    #f['layout']['xaxis3']['domain'] = [0.98, 1.]\n",
    "\n",
    "    f.layout.hovermode = 'closest'\n",
    "\n",
    "\n",
    "    # Set background color to zero\n",
    "    fig.layout.template = \"plotly_dark\"\n",
    "    fig.layout.plot_bgcolor = \"rgba(0,0,0,0)\"\n",
    "    fig.layout.paper_bgcolor = \"rgba(0,0,0,0)\"\n",
    "\n",
    "\n",
    "    # create our callback function\n",
    "    debug_view = widgets.Output(layout={'border': '1px solid black'})\n",
    "    @debug_view.capture(clear_output=True)\n",
    "    def update(trace, points, selector):\n",
    "        #global p2\n",
    "        #global p3\n",
    "        for i in points.point_inds:\n",
    "\n",
    "            index_init = i\n",
    "            array_init_genes, l_genes_init = zip(*sorted(zip(array_exp_genes[index_init:index_init+1,:].T, l_genes), reverse=True))\n",
    "            array_init_genes = np.array(array_init_genes)[:n_to_keep]\n",
    "            l_genes_init = l_genes_init[:n_to_keep]\n",
    "            array_init_lipids, l_lipids_init = zip(*sorted(zip(array_exp_lipids[index_init:index_init+1,:].T, l_name_lipids), reverse=True))\n",
    "            array_init_lipids = np.array(array_init_lipids)[:n_to_keep]\n",
    "            l_lipids_init = l_lipids_init[:n_to_keep]\n",
    "\n",
    "            with f.batch_update():\n",
    "\n",
    "                p2.y = l_genes_init\n",
    "                p2.z = array_init_genes\n",
    "                p3.y = l_lipids_init\n",
    "                p3.z = array_init_lipids\n",
    "\n",
    "    p1.on_click(update)\n",
    "f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if plot_widget:\n",
    "    debug_view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cluster pixel according to genes and lipids and compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(array_exp_genes.shape, array_exp_lipids.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute a UMAP projection and do clustering on the result\n",
    "umap_2d_lipids = UMAP(n_components=2, init='random', random_state=0)\n",
    "proj_2d_lipids = umap_2d_lipids.fit_transform(array_exp_lipids)\n",
    "kmeans_lipids = KMeans(n_clusters=100, random_state=0).fit(proj_2d_lipids)\n",
    "\n",
    "y = pdist(kmeans_lipids.cluster_centers_)\n",
    "Z = ward(y)\n",
    "leaves_ordered = list(leaves_list(optimal_leaf_ordering(Z, y)))\n",
    "\n",
    "l_colors_lipids = []\n",
    "for label in kmeans_lipids.labels_:\n",
    "    l_colors_lipids.append(leaves_ordered.index(label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = go.Figure(data=go.Scattergl(\n",
    "    x = proj_2d_lipids[:,0],\n",
    "    y = proj_2d_lipids[:,1],\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        color = l_colors_lipids,\n",
    "        size=4,\n",
    "        opacity = 0.4,\n",
    "        colorscale = px.colors.sequential.Jet,\n",
    "    )\n",
    "))\n",
    "\n",
    "\n",
    "fig.update_traces(textposition='top center', textfont_size=8, showlegend=False)\n",
    "fig.update_layout(\n",
    "    width=int(800),\n",
    "    height=int(800),\n",
    "    font_size=7,\n",
    "    title_font_size=12,\n",
    "    title={\n",
    "        'text' : 'UMAP and K-Means voxel clustering using lipidomics (MALDI) data',\n",
    "        'y':0.92,\n",
    "        'x':0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top'},\n",
    "    template='simple_white'\n",
    ")\n",
    "fig.show()\n",
    "fig.write_image(\"notebooks/scRNAseq/umap1.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute a UMAP projection and do clustering on the result\n",
    "umap_2d_genes = UMAP(n_components=2, init='random', random_state=0)\n",
    "proj_2d_genes = umap_2d_genes.fit_transform(array_exp_genes)\n",
    "kmeans_genes = KMeans(n_clusters=10, random_state=0).fit(proj_2d_genes)\n",
    "\n",
    "y = pdist(kmeans_genes.cluster_centers_)\n",
    "Z = ward(y)\n",
    "leaves_ordered = list(leaves_list(optimal_leaf_ordering(Z, y)))\n",
    "\n",
    "l_colors_genes = []\n",
    "for label in kmeans_genes.labels_:\n",
    "    l_colors_genes.append(leaves_ordered.index(label))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig = go.Figure(data=go.Scattergl(\n",
    "    x = proj_2d_genes[:,0],\n",
    "    y = proj_2d_genes[:,1],\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        color = l_colors_genes,\n",
    "        size=4,\n",
    "        opacity = 0.4,\n",
    "        colorscale = px.colors.sequential.Jet,\n",
    "    )\n",
    "))\n",
    "\n",
    "\n",
    "fig.update_traces(textposition='top center', textfont_size=8, showlegend=False)\n",
    "fig.update_layout(\n",
    "    width=int(800),\n",
    "    height=int(800),\n",
    "    font_size=7,\n",
    "    title_font_size=12,\n",
    "    title={\n",
    "        'text' : 'UMAP and K-Means voxel clustering using gene (scRNAseq) data',\n",
    "        'y':0.92,\n",
    "        'x':0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top'},\n",
    "    template='simple_white'\n",
    ")\n",
    "fig.show()\n",
    "fig.write_image(\"notebooks/scRNAseq/umap2.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = go.Figure(data=go.Scattergl(\n",
    "    x = proj_2d_genes[:,0],\n",
    "    y = proj_2d_genes[:,1],\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        color = l_colors_lipids,\n",
    "        size=4,\n",
    "        opacity = 0.4,\n",
    "        colorscale = px.colors.sequential.Jet,\n",
    "    )\n",
    "))\n",
    "\n",
    "\n",
    "fig.update_traces(textposition='top center', textfont_size=8, showlegend=False)\n",
    "fig.update_layout(\n",
    "    width=int(800),\n",
    "    height=int(800),\n",
    "    font_size=7,\n",
    "    title_font_size=12,\n",
    "    title={\n",
    "        'text' : 'UMAP voxel gene data (scRNAseq) projection using Lipid data (MALDI) for clustering',\n",
    "        'y':0.92,\n",
    "        'x':0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top'},\n",
    "    template='simple_white'\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "fig.write_image(\"notebooks/scRNAseq/umap3.pdf\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = go.Figure(data=go.Scattergl(\n",
    "    x = proj_2d_lipids[:,0],\n",
    "    y = proj_2d_lipids[:,1],\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        color = l_colors_genes,\n",
    "        size=4,\n",
    "        opacity = 0.4,\n",
    "        colorscale = px.colors.sequential.Jet,\n",
    "    )\n",
    "))\n",
    "\n",
    "\n",
    "fig.update_traces(textposition='top center', textfont_size=8, showlegend=False)\n",
    "fig.update_layout(\n",
    "    width=int(800),\n",
    "    height=int(800),\n",
    "    font_size=7,\n",
    "    title_font_size=12,\n",
    "    title={\n",
    "        'text' : 'UMAP lipid data (MALDI) projection using gene data (scRNAseq) for clustering',\n",
    "        'y':0.92,\n",
    "        'x':0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top'},\n",
    "    template='simple_white'\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "fig.write_image(\"notebooks/scRNAseq/umap4.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project back cluster in brain space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if True:\n",
    "    scatter = go.Scatter3d(\n",
    "        x=xmol,\n",
    "        y=zmol,\n",
    "        z=-ymol,\n",
    "        mode='markers',\n",
    "        marker=dict(\n",
    "            size=1,\n",
    "            color=l_colors_lipids,\n",
    "            opacity=1.,\n",
    "            colorscale = px.colors.sequential.Jet,\n",
    "        ),\n",
    "    )\n",
    "    scatter['hoverinfo']= \"all\"\n",
    "\n",
    "\n",
    "\n",
    "    # Get root figure\n",
    "    root_data = figures._storage.return_shelved_object(\n",
    "        \"figures/3D_page\",\n",
    "        \"volume_root\",\n",
    "        force_update=False,\n",
    "        compute_function=figures.compute_3D_root_volume,\n",
    "    )\n",
    "\n",
    "    root_data_y = copy.deepcopy(root_data['y'])\n",
    "    root_data_z = copy.deepcopy(root_data['z'])\n",
    "    root_data['y'] = root_data_z \n",
    "    root_data['z'] = -root_data_y\n",
    "    root_data['hoverinfo']= 'skip'\n",
    "\n",
    "\n",
    "\n",
    "    fig = go.Figure(data=[root_data, scatter])\n",
    "\n",
    "\n",
    "    # Hide grey background\n",
    "    fig.update_layout(\n",
    "        margin=dict(t=0, r=0, b=0, l=0),\n",
    "            scene=dict(\n",
    "                xaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"rgba(0,0,0,0)\", gridcolor=\"rgba(0,0,0,0)\", showticklabels=False, showgrid= False, zeroline = False, showline = False,),\n",
    "                yaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"rgba(0,0,0,0)\", gridcolor=\"rgba(0,0,0,0)\", showticklabels=False, showgrid= False, zeroline = False, showline = False,),\n",
    "                zaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"rgba(0,0,0,0)\", gridcolor=\"rgba(0,0,0,0)\", showticklabels=False, showgrid= False, zeroline = False, showline = False,),\n",
    "            ),\n",
    "            width=1000,\n",
    "            height=600,\n",
    "            title={\n",
    "                'y':0.9,\n",
    "                'x':0.5,\n",
    "                'xanchor': 'center',\n",
    "                'text': 'Lipids clusters (MALDI data)',\n",
    "                'yanchor': 'top'}\n",
    "    )\n",
    "\n",
    "    # Set background color to zero\n",
    "    fig.layout.template = \"plotly_dark\"\n",
    "    fig.layout.plot_bgcolor = \"rgba(0,0,0,0)\"\n",
    "    fig.layout.paper_bgcolor = \"rgba(0,0,0,0)\"\n",
    "\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if False:\n",
    "    scatter = go.Scatter3d(\n",
    "        x=x_mol,\n",
    "        y=z_mol,\n",
    "        z=-y_mol,\n",
    "        mode='markers',\n",
    "        marker=dict(\n",
    "            size=1,\n",
    "            color=kmeans_genes.labels_,    \n",
    "            opacity=1.,\n",
    "            colorscale = px.colors.qualitative.T10,\n",
    "        ),\n",
    "    )\n",
    "    scatter['hoverinfo']= \"all\"\n",
    "\n",
    "    # Get root figure\n",
    "    root_data = figures._storage.return_shelved_object(\n",
    "        \"figures/3D_page\",\n",
    "        \"volume_root\",\n",
    "        force_update=False,\n",
    "        compute_function=figures.compute_3D_root_volume,\n",
    "    )\n",
    "\n",
    "    root_data_y = copy.deepcopy(root_data['y'])\n",
    "    root_data_z = copy.deepcopy(root_data['z'])\n",
    "    root_data['y'] = root_data_z \n",
    "    root_data['z'] = -root_data_y\n",
    "\n",
    "\n",
    "    fig = go.Figure(data=[root_data, scatter])\n",
    "\n",
    "\n",
    "    # Hide grey background\n",
    "    fig.update_layout(\n",
    "        margin=dict(t=0, r=0, b=0, l=0),\n",
    "            scene=dict(\n",
    "                xaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"rgba(0,0,0,0)\", gridcolor=\"rgba(0,0,0,0)\", showticklabels=False, showgrid= False, zeroline = False, showline = False,),\n",
    "                yaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"rgba(0,0,0,0)\", gridcolor=\"rgba(0,0,0,0)\", showticklabels=False, showgrid= False, zeroline = False, showline = False,),\n",
    "                zaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"rgba(0,0,0,0)\", gridcolor=\"rgba(0,0,0,0)\", showticklabels=False, showgrid= False, zeroline = False, showline = False,),\n",
    "            ),\n",
    "            width=1000,\n",
    "            height=600,\n",
    "            title={\n",
    "                'y':0.9,\n",
    "                'x':0.5,\n",
    "                'xanchor': 'center',\n",
    "                'text': 'Genes clusters (scRNAseq data)',\n",
    "                'yanchor': 'top'},\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "    # Set background color to zero\n",
    "    fig.layout.template = \"plotly_dark\"\n",
    "    fig.layout.plot_bgcolor = \"rgba(0,0,0,0)\"\n",
    "    fig.layout.paper_bgcolor = \"rgba(0,0,0,0)\"\n",
    "\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Function for 3D rotation\n",
    "def rotate_y(x, y, z, theta):\n",
    "    w = x+1j*z\n",
    "    return np.real(np.exp(1j*theta)*w), y, np.imag(np.exp(1j*theta)*w)\n",
    "    \n",
    "def animate_figure(fig, output = 'output.gif', multiprocessing = True):\n",
    "    # Initial coordinates\n",
    "    x_eye=1.25*0.85\n",
    "    y_eye=0.1*0.85\n",
    "    z_eye=1.25*0.85\n",
    "\n",
    "    # Update with rotation\n",
    "    time_domain = np.arange(0, 500, 1)\n",
    "    l_fig = []\n",
    "    for image, t in enumerate(time_domain):\n",
    "        fig_rotated = copy.deepcopy(fig)\n",
    "\n",
    "        # Rotate camera\n",
    "        xe, ye, ze = rotate_y(x_eye, y_eye, z_eye, -t/40)\n",
    "        camera = dict(\n",
    "            up=dict(x=0, y=0, z=1),\n",
    "            center=dict(x=0, y=0, z=0),\n",
    "            eye=dict(x=xe, y=ze, z=ye),\n",
    "        )\n",
    "        fig_rotated.update_layout(scene_camera=camera)\n",
    "\n",
    "        # Append final figure to the list of images\n",
    "        l_fig.append((image,fig_rotated))\n",
    "    \n",
    "\n",
    "    def write_figure(t_fig):\n",
    "        image,fig = t_fig\n",
    "        fig.write_image(\"notebooks/scRNAseq/temp/\"+ format(image,'04d')+\".jpg\")\n",
    "        \n",
    "    if multiprocessing:\n",
    "        with Pool(processes=6) as pool:\n",
    "            [x for x in pool.map(write_figure, l_fig)]\n",
    "            \n",
    "    else:\n",
    "        for image,fig in l_fig:\n",
    "            print(\"Plotting figure \" + str(image))\n",
    "            fig.write_image(\"notebooks/scRNAseq/temp/\"+ format(image,'04d')+\".jpg\")\n",
    "            \n",
    "    # Merge images into a gif\n",
    "    i = \"notebooks/scRNAseq/temp/*.jpg\"\n",
    "    o = \"notebooks/scRNAseq/temp/\" + output\n",
    "    subprocess.call(\"convert -delay 2 -loop 0 \" + i + \" \" + o, shell=True)\n",
    "    \n",
    "animate_figure(fig, output = 'output_progressive_clustering_lipids.gif', multiprocessing = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Animation of the most different lipids in 3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "decrease_resolution_factor = 7\n",
    "ll_exp_lipids = []\n",
    "l_name_lipids = []\n",
    "if True:\n",
    "\n",
    "    # Simulate a click on all lipid names\n",
    "    for name in sorted(\n",
    "        figures._data.get_annotations_MAIA_transformed_lipids(brain_1=True).name.unique()\n",
    "    ):\n",
    "        structures = figures._data.get_annotations_MAIA_transformed_lipids(brain_1=True)[\n",
    "            figures._data.get_annotations_MAIA_transformed_lipids(brain_1=True)[\"name\"] == name\n",
    "        ].structure.unique()\n",
    "        for structure in sorted(structures):\n",
    "            cations = figures._data.get_annotations_MAIA_transformed_lipids(brain_1=True)[\n",
    "                (\n",
    "                    figures._data.get_annotations_MAIA_transformed_lipids(brain_1=True)[\"name\"]\n",
    "                    == name\n",
    "                )\n",
    "                & (\n",
    "                    figures._data.get_annotations_MAIA_transformed_lipids(brain_1=True)[\n",
    "                        \"structure\"\n",
    "                    ]\n",
    "                    == structure\n",
    "                )\n",
    "            ].cation.unique()\n",
    "            for cation in sorted(cations):\n",
    "                l_selected_lipids = []\n",
    "                for slice_index in range(figures._data.get_slice_number()):\n",
    "\n",
    "                    # Find lipid location\n",
    "                    l_lipid_loc = (\n",
    "                        figures._data.get_annotations()\n",
    "                        .index[\n",
    "                            (figures._data.get_annotations()[\"name\"] == name)\n",
    "                            & (figures._data.get_annotations()[\"structure\"] == structure)\n",
    "                            & (figures._data.get_annotations()[\"slice\"] == slice_index)\n",
    "                            & (figures._data.get_annotations()[\"cation\"] == cation)\n",
    "                        ]\n",
    "                        .tolist()\n",
    "                    )\n",
    "\n",
    "                    # If several lipids correspond to the selection, we have a problem...\n",
    "                    if len(l_lipid_loc) > 1:\n",
    "                        logging.warning(\"More than one lipid corresponds to the selection\")\n",
    "                        l_lipid_loc = [l_lipid_loc[-1]]\n",
    "                    # If no lipid correspond to the selection, set to -1\n",
    "                    if len(l_lipid_loc) == 0:\n",
    "                        l_lipid_loc = [-1]\n",
    "\n",
    "                    # add lipid index for each slice\n",
    "                    l_selected_lipids.append(l_lipid_loc[0])\n",
    "\n",
    "                # Get final lipid name\n",
    "                lipid_string = name + \" \" + structure + \" \" + cation\n",
    "\n",
    "                # If lipid is present in at least one slice\n",
    "                if np.sum(l_selected_lipids) > -figures._data.get_slice_number():\n",
    "\n",
    "                    # Build the list of mz boundaries for each peak and each index\n",
    "                    lll_lipid_bounds = [\n",
    "                        [\n",
    "                            [\n",
    "                                (\n",
    "                                    float(figures._data.get_annotations().iloc[index][\"min\"]),\n",
    "                                    float(figures._data.get_annotations().iloc[index][\"max\"]),\n",
    "                                )\n",
    "                            ]\n",
    "                            if index != -1\n",
    "                            else None\n",
    "                            for index in [lipid_1_index, -1, -1]\n",
    "                        ]\n",
    "                        for lipid_1_index in l_selected_lipids\n",
    "                    ]\n",
    "                    print(\"getting data for lipid \", lipid_string)\n",
    "                    \n",
    "                    l_expr = []\n",
    "\n",
    "                    try:\n",
    "                        interpolated_array = figures.compute_3D_volume_figure(ll_t_bounds = lll_lipid_bounds, name_lipid_1=lipid_string,decrease_dimensionality_factor=decrease_resolution_factor,return_interpolated_array=True, structure_guided_interpolation=True, divider_radius=12)\n",
    "                    except Exception as e:\n",
    "                        print(e)\n",
    "                    \n",
    "                    l_name_lipids.append(lipid_string)\n",
    "                    ll_exp_lipids.append(interpolated_array)\n",
    "\n",
    "\n",
    "\n",
    "    array_exp_lipids_for_display = np.array(ll_exp_lipids)\n",
    "    \n",
    "    with open('notebooks/scRNAseq/data/array_exp_lipids_for_display.npy', 'wb') as f:\n",
    "        np.save(f, array_exp_lipids_for_display)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(array_exp_lipids_for_display.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_array_exp_lipids_for_display_flat = []\n",
    "for arr in array_exp_lipids_for_display:\n",
    "    l_array_exp_lipids_for_display_flat.append(arr.flatten())\n",
    "array_exp_lipids_for_display_flat = np.array(l_array_exp_lipids_for_display_flat)\n",
    "l_array_exp_lipids_for_display_flat = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute a UMAP projection and do clustering on the result\n",
    "umap_2d_all_lipids = UMAP(n_components=2, init='random', random_state=0)\n",
    "proj_2d_all_lipids = umap_2d_all_lipids.fit_transform(array_exp_lipids_for_display_flat)\n",
    "kmeans_all_lipids = KMeans(n_clusters=len(proj_2d_all_lipids), random_state=0).fit(proj_2d_all_lipids)\n",
    "\n",
    "y = pdist(kmeans_all_lipids.cluster_centers_)\n",
    "Z = ward(y)\n",
    "leaves_ordered = list(leaves_list(optimal_leaf_ordering(Z, y)))\n",
    "\n",
    "l_colors_all_lipids = []\n",
    "for label in kmeans_all_lipids.labels_:\n",
    "    l_colors_all_lipids.append(leaves_ordered.index(label))\n",
    "    \n",
    "    \n",
    "fig = go.Figure(data=go.Scattergl(\n",
    "    x = proj_2d_all_lipids[:,0],\n",
    "    y = proj_2d_all_lipids[:,1],\n",
    "    mode='markers',\n",
    "    text = [str(i) + '\\n\\n' + x for i, x in enumerate(l_name_lipids)],\n",
    "    marker=dict(\n",
    "        color = l_colors_all_lipids,\n",
    "        size=4,\n",
    "        opacity = 0.4,\n",
    "        colorscale = px.colors.sequential.Jet,\n",
    "    )\n",
    "))\n",
    "\n",
    "\n",
    "fig.update_traces(textposition='top center', textfont_size=8, showlegend=False)\n",
    "fig.update_layout(\n",
    "    width=int(800),\n",
    "    height=int(800),\n",
    "    font_size=7,\n",
    "    title_font_size=12,\n",
    "    title={\n",
    "        'text' : 'UMAP and K-Means voxel clustering using lipidomics (MALDI) data',\n",
    "        'y':0.92,\n",
    "        'x':0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top'},\n",
    "    template='simple_white'\n",
    ")\n",
    "fig.show()\n",
    "fig.write_image(\"notebooks/scRNAseq/umap_all_lipids.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@njit\n",
    "def median_symmetric_filter(array_lipid):\n",
    "    size_filter = 1\n",
    "    array_lipid_filtered = np.zeros_like(array_lipid)\n",
    "    for slice_index in range(int(array_lipid.shape[2]/2)+1):\n",
    "        for x in range(size_filter,array_lipid.shape[0]-size_filter):\n",
    "            for y in range(size_filter,array_lipid.shape[1]-size_filter):\n",
    "                # Compute median in a square of size 2*size_filter+1\n",
    "                median = np.median(np.vstack((array_lipid[x-size_filter:x+size_filter+1, y-size_filter:y+size_filter+1, slice_index], array_lipid[x-size_filter:x+size_filter+1, y-size_filter:y+size_filter+1, -slice_index])))\n",
    "                array_lipid_filtered[x,y,slice_index] = median\n",
    "                array_lipid_filtered[x,y,-slice_index] = median\n",
    "                \n",
    "    if array_lipid.shape[2]%2!=0:\n",
    "        array_lipid_filtered[:,:,int(array_lipid.shape[2]/2)+1] = array_lipid[:,:,int(array_lipid.shape[2]/2)+1]\n",
    "    return array_lipid_filtered\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get lipid indexes to plot\n",
    "lipid_1 = 36\n",
    "lipid_2 = 25\n",
    "lipid_3 = 50\n",
    "\n",
    "# Compute a median symmetric filter\n",
    "array_lipid_1 = median_symmetric_filter(array_exp_lipids_for_display[lipid_1])\n",
    "array_lipid_2 = median_symmetric_filter(array_exp_lipids_for_display[lipid_2])\n",
    "array_lipid_3 = median_symmetric_filter(array_exp_lipids_for_display[lipid_3])\n",
    "\n",
    "# Initial coordinates\n",
    "x_eye=1.25*0.85\n",
    "y_eye=0.1*0.85\n",
    "z_eye=1.25*0.85\n",
    "    \n",
    "\n",
    "# Get the corresponding coordinates\n",
    "shape_x, shape_y, shape_z = array_exp_lipids_for_display.shape[1:]\n",
    "X, Y, Z = np.mgrid[\n",
    "    0 : shape_x\n",
    "    / 1000\n",
    "    * 25\n",
    "    * decrease_resolution_factor : shape_x\n",
    "    * 1j,\n",
    "    0 : shape_y\n",
    "    / 1000\n",
    "    * 25\n",
    "    * decrease_resolution_factor : shape_y\n",
    "    * 1j,\n",
    "    0 : shape_z\n",
    "    / 1000\n",
    "    * 25\n",
    "    * decrease_resolution_factor : shape_z\n",
    "    * 1j,\n",
    "]\n",
    "\n",
    "if True:\n",
    "    \n",
    "    # Compute first maldi lipid in 3D\n",
    "    volume_maldi = go.Volume(\n",
    "        x=X.flatten(),\n",
    "        y=Z.flatten(),\n",
    "        z=-Y.flatten(),\n",
    "        value=array_lipid_1.flatten(),\n",
    "        isomin=0.01,\n",
    "        isomax=1.5,\n",
    "        opacity = 0,\n",
    "        opacityscale='uniform',\n",
    "        surface_count=10,  # needs to be a large number for good volume rendering\n",
    "        colorscale=\"Reds\",\n",
    "    )\n",
    "\n",
    "    # Compute second maldi lipid in 3D\n",
    "    volume_maldi_2 = go.Volume(\n",
    "        x=X.flatten(),\n",
    "        y=Z.flatten(),\n",
    "        z=-Y.flatten(),\n",
    "        value=array_lipid_2.flatten(),\n",
    "        isomin=0.01,\n",
    "        isomax=1.5,\n",
    "        opacity = 0,\n",
    "        opacityscale='uniform',\n",
    "        surface_count=10,  # needs to be a large number for good volume rendering\n",
    "        colorscale=\"Blues\",\n",
    "    )\n",
    "    \n",
    "    # Compute thid maldi lipid in 3D\n",
    "    volume_maldi_3 = go.Volume(\n",
    "        x=X.flatten(),\n",
    "        y=Z.flatten(),\n",
    "        z=-Y.flatten(),\n",
    "        value=array_lipid_3.flatten(),\n",
    "        isomin=0.01,\n",
    "        isomax=1.5,\n",
    "        opacity = 0,\n",
    "        opacityscale='uniform',\n",
    "        surface_count=10,  # needs to be a large number for good volume rendering\n",
    "        colorscale=\"Greens\",\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "    # Get root figure\n",
    "    root_data = figures._storage.return_shelved_object(\n",
    "        \"figures/3D_page\",\n",
    "        \"volume_root\",\n",
    "        force_update=False,\n",
    "        compute_function=figures.compute_3D_root_volume,\n",
    "    )\n",
    "\n",
    "    root_data_y = copy.deepcopy(root_data['y'])\n",
    "    root_data_z = copy.deepcopy(root_data['z'])\n",
    "    root_data['y'] = root_data_z \n",
    "    root_data['z'] = -root_data_y\n",
    "    root_data['hoverinfo']= 'skip'\n",
    "    \n",
    "    \n",
    "\n",
    "    # Define figure and layout\n",
    "    fig = go.Figure(data=[root_data, volume_maldi, volume_maldi_2, volume_maldi_3])\n",
    "\n",
    "    # Transparent background\n",
    "    fig.update_layout(\n",
    "        margin=dict(t=0, r=0, b=0, l=0),\n",
    "        scene=dict(\n",
    "            xaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"white\", gridcolor=\"white\", showticklabels=False),\n",
    "            yaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"white\", gridcolor=\"white\", showticklabels=False),\n",
    "            zaxis=dict(backgroundcolor=\"rgba(0,0,0,0)\",color=\"white\", gridcolor=\"white\", showticklabels=False),\n",
    "        ),\n",
    "        # Bigger size for nicer display\n",
    "        width=1000,\n",
    "        height=600,\n",
    "    )\n",
    "    fig.update_layout(\n",
    "        title={\n",
    "            'y':0.94,\n",
    "            'x':0.5,\n",
    "            'xanchor': 'center',\n",
    "            'yanchor': 'top'}\n",
    "    )\n",
    "\n",
    "    # No colorbar\n",
    "    fig.update_traces(showscale=False)\n",
    "\n",
    "    #fig.show()\n",
    "    #fig.write_image(\"notebooks/scRNAseq/temp/test.jpg\")\n",
    "    \n",
    "    if True:\n",
    "        # Update with rotation\n",
    "        t1 = 0\n",
    "        t2 = 200\n",
    "        t3 = 250\n",
    "        t4 = 450\n",
    "        t5 = 500\n",
    "        t6 = 700\n",
    "        time_domain = np.arange(t1, t4, 1)\n",
    "\n",
    "\n",
    "        l_fig = []\n",
    "        for image, t in enumerate(time_domain):\n",
    "            fig_rotated = copy.deepcopy(fig)\n",
    "\n",
    "            # Remove reference always after t=150, and display data always\n",
    "            if t>=t1 and t<t2:\n",
    "                # Title\n",
    "                fig_rotated.update_layout(title_text=l_name_lipids[lipid_1])\n",
    "                # First lipid is displayed fully\n",
    "                fig_rotated.data[1].opacity = 1\n",
    "\n",
    "            # Switch between two lipids\n",
    "            elif t>=t2 and t<t3:\n",
    "                # No more title as switch taking place\n",
    "                fig_rotated.update_layout(title_text='')\n",
    "\n",
    "                # Fade away lipid 1\n",
    "                fig_rotated.data[1].opacity = (t3-t)/(t3-t2)\n",
    "                # Progressivelu show lipid 2\n",
    "                fig_rotated.data[2].opacity = (t-t2)/(t3-t2)\n",
    "\n",
    "            # Just leave lipid 2\n",
    "            elif t>=t3:# and t<t4:\n",
    "                # Title\n",
    "                fig_rotated.update_layout(title_text=l_name_lipids[lipid_2])\n",
    "                # Lipid 2 is displayed fully\n",
    "                fig_rotated.data[2].opacity = 1\n",
    "\n",
    "#            # Progressively display lipid 3\n",
    "#            elif t>=t4 and t<t5:\n",
    "#                # No more title as switch taking place\n",
    "#                fig_rotated.update_layout(title_text='')\n",
    "#                # Lipid 2 fades away\n",
    "#                fig_rotated.data[2].opacity = (t5-t)/(t5-t4)\n",
    "#                # Lipid 3 appears\n",
    "#                fig_rotated.data[3].opacity = (t-t4)/(t5-t4)\n",
    "#\n",
    "#            # Just leave new lipid\n",
    "#            elif t>=t5:\n",
    "#                # Title\n",
    "#                fig_rotated.update_layout(title_text=l_name_lipids[lipid_3])\n",
    "#\n",
    "#                # Lipid 3 is displayed fully\n",
    "#                fig_rotated.data[3].opacity = 1\n",
    "\n",
    "            # Rotate camera\n",
    "            xe, ye, ze = rotate_y(x_eye, y_eye, z_eye, -t/40)\n",
    "            camera = dict(\n",
    "                up=dict(x=0, y=0, z=1),\n",
    "                center=dict(x=0, y=0, z=0),\n",
    "                eye=dict(x=xe, y=ze, z=ye),\n",
    "            )\n",
    "            fig_rotated.update_layout(scene_camera=camera)\n",
    "\n",
    "            # Append final figure to the list of images\n",
    "            l_fig.append((image,fig_rotated))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output = 'two_lipids.gif'\n",
    "\n",
    "def write_figure(t_fig):\n",
    "    image,fig = t_fig\n",
    "    fig.write_image(\"notebooks/scRNAseq/temp/\"+ format(image,'04d')+\".jpg\")\n",
    "\n",
    "multiprocessing = True\n",
    "if multiprocessing:\n",
    "    with Pool(processes=10) as pool:\n",
    "        print(\"Wallah\")\n",
    "        [x for x in pool.map(write_figure, l_fig)]\n",
    "\n",
    "else:\n",
    "    for image,fig in l_fig:\n",
    "        print(\"Plotting figure \" + str(image))\n",
    "        fig.write_image(\"notebooks/scRNAseq/temp/\"+ + format(image,'04d')+\".jpg\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge images into a gif\n",
    "i = \"notebooks/scRNAseq/temp/*.jpg\"\n",
    "o = \"notebooks/scRNAseq/temp/\" + output\n",
    "subprocess.call(\"convert -delay 3 -loop 0 \" + i + \" \" + o, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "0c1aa729cc35b9a783763c24c4069d7da678acf641f89d4e1df25bf02079ad65"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
