{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook to analyze and display cell type data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load important modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.git', '.gitattributes', '.vscode', 'LICENSE.md', 'TODO.py', 'app.py', 'assets', 'config.py', 'data_sample', 'documentation', 'index.py', 'js', 'main.py', 'modules', 'notebooks', 'pages', 'readme', 'style', 'data', 'README.md', '.ipynb_checkpoints', '.gitignore', 'Dockerfile', 'nohup.out', '.mypy_cache', 'requirements.txt', '__pycache__']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<threadpoolctl.threadpool_limits at 0x7ff65811fb50>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standard modules\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd \n",
    "import plotly.graph_objects as go\n",
    "import matplotlib.pyplot as plt\n",
    "import logging\n",
    "from numba import njit\n",
    "from sklearn import linear_model\n",
    "\n",
    "\n",
    "# Move to root directory for easier module handling\n",
    "os.chdir(\"../..\")\n",
    "print(os.listdir(\".\"))\n",
    "\n",
    "#LBAE imports\n",
    "from modules.maldi_data import MaldiData\n",
    "from modules.figures import Figures\n",
    "from modules.atlas import Atlas\n",
    "from modules.storage import Storage\n",
    "from modules.scRNAseq import ScRNAseq\n",
    "\n",
    "# multithreading/multiprocessing\n",
    "from multiprocessing import Pool\n",
    "import multiprocessing\n",
    "from threadpoolctl import threadpool_limits\n",
    "\n",
    "# set thread limit\n",
    "threadpool_limits(16)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load LBAE objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_data = \"data/whole_dataset/\"\n",
    "path_annotations = \"data/annotations/\"\n",
    "path_db = \"data/app_data/data.db\"\n",
    "\n",
    "# Load shelve database\n",
    "storage = Storage(path_db)\n",
    "\n",
    "# Load data\n",
    "data = MaldiData(path_data, path_annotations)\n",
    "\n",
    "# Load Atlas and Figures objects. At first launch, many objects will be precomputed and shelved in\n",
    "# the classes Atlas and Figures.\n",
    "atlas = Atlas(data, storage, resolution=25, sample = True)\n",
    "scRNAseq = ScRNAseq()\n",
    "figures = Figures(data, storage, atlas, scRNAseq, sample = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load cell type data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Regions</th>\n",
       "      <th>Excitatory [mm-3]</th>\n",
       "      <th>Inhibitory [mm-3]</th>\n",
       "      <th>Modulatory [mm-3]</th>\n",
       "      <th>Astrocytes [mm-3]</th>\n",
       "      <th>Oligodendrocytes [mm-3]</th>\n",
       "      <th>Microglia [mm-3]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>root</td>\n",
       "      <td>137352.601147</td>\n",
       "      <td>11751.903955</td>\n",
       "      <td>849.961392</td>\n",
       "      <td>12782.383276</td>\n",
       "      <td>43770.289886</td>\n",
       "      <td>25613.012771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Basic cell groups and regions</td>\n",
       "      <td>153231.404562</td>\n",
       "      <td>13110.496156</td>\n",
       "      <td>948.222144</td>\n",
       "      <td>12058.873983</td>\n",
       "      <td>38918.967718</td>\n",
       "      <td>24809.444601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cerebrum</td>\n",
       "      <td>79759.103521</td>\n",
       "      <td>14290.585238</td>\n",
       "      <td>514.529938</td>\n",
       "      <td>10005.391872</td>\n",
       "      <td>32256.018628</td>\n",
       "      <td>23460.962106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cerebral cortex</td>\n",
       "      <td>90968.098789</td>\n",
       "      <td>10860.004085</td>\n",
       "      <td>492.496350</td>\n",
       "      <td>9855.058423</td>\n",
       "      <td>31800.904552</td>\n",
       "      <td>26231.659016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cortical plate</td>\n",
       "      <td>91783.848364</td>\n",
       "      <td>10988.904301</td>\n",
       "      <td>487.060649</td>\n",
       "      <td>9882.181587</td>\n",
       "      <td>31371.949122</td>\n",
       "      <td>26817.575208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>mammillary peduncle</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21523.131673</td>\n",
       "      <td>93115.065243</td>\n",
       "      <td>683.274021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>951</th>\n",
       "      <td>epithalamus related</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21679.418886</td>\n",
       "      <td>74844.640106</td>\n",
       "      <td>44423.860885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>stria medullaris</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26794.563438</td>\n",
       "      <td>80137.764179</td>\n",
       "      <td>48436.063474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>fasciculus retroflexus</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9203.914308</td>\n",
       "      <td>63826.500926</td>\n",
       "      <td>36811.425549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>habenular commissure</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>42094.545455</td>\n",
       "      <td>82967.272727</td>\n",
       "      <td>45498.181818</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>955 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Regions  Excitatory [mm-3]  Inhibitory [mm-3]  \\\n",
       "0                             root      137352.601147       11751.903955   \n",
       "1    Basic cell groups and regions      153231.404562       13110.496156   \n",
       "2                         Cerebrum       79759.103521       14290.585238   \n",
       "3                  Cerebral cortex       90968.098789       10860.004085   \n",
       "4                   Cortical plate       91783.848364       10988.904301   \n",
       "..                             ...                ...                ...   \n",
       "950            mammillary peduncle           0.000000           0.000000   \n",
       "951            epithalamus related           0.000000           0.000000   \n",
       "952               stria medullaris           0.000000           0.000000   \n",
       "953         fasciculus retroflexus           0.000000           0.000000   \n",
       "954           habenular commissure           0.000000           0.000000   \n",
       "\n",
       "     Modulatory [mm-3]  Astrocytes [mm-3]  Oligodendrocytes [mm-3]  \\\n",
       "0           849.961392       12782.383276             43770.289886   \n",
       "1           948.222144       12058.873983             38918.967718   \n",
       "2           514.529938       10005.391872             32256.018628   \n",
       "3           492.496350        9855.058423             31800.904552   \n",
       "4           487.060649        9882.181587             31371.949122   \n",
       "..                 ...                ...                      ...   \n",
       "950           0.000000       21523.131673             93115.065243   \n",
       "951           0.000000       21679.418886             74844.640106   \n",
       "952           0.000000       26794.563438             80137.764179   \n",
       "953           0.000000        9203.914308             63826.500926   \n",
       "954           0.000000       42094.545455             82967.272727   \n",
       "\n",
       "     Microglia [mm-3]  \n",
       "0        25613.012771  \n",
       "1        24809.444601  \n",
       "2        23460.962106  \n",
       "3        26231.659016  \n",
       "4        26817.575208  \n",
       "..                ...  \n",
       "950        683.274021  \n",
       "951      44423.860885  \n",
       "952      48436.063474  \n",
       "953      36811.425549  \n",
       "954      45498.181818  \n",
       "\n",
       "[955 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_cells = pd.read_csv('notebooks/cell_type_atlas/data/Data_Sheet_2_A Cell Atlas for the Mouse Brain.csv', usecols=[0,4,5,6,7,8,9] )\n",
    "table_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of leafs in the hierarchy of structures\n",
    "l_leafs_ids = []\n",
    "l_leafs_acronyms = []\n",
    "l_leafs_names = []\n",
    "\n",
    "for x in atlas.dic_acronym_children_id:\n",
    "    if len(atlas.dic_acronym_children_id[x]) == 1:\n",
    "        l_leafs_ids.extend(list(atlas.dic_acronym_children_id[x]))\n",
    "        l_leafs_acronyms.append(x)\n",
    "        l_leafs_names.append(atlas.dic_acronym_name[x])\n",
    "    if len(atlas.dic_acronym_children_id[x]) == 0:\n",
    "        raise ValueError(\"No leaf found for structure: \" + x)\n",
    "\n",
    "dic_names_id = {x:y for x,y in zip(l_leafs_names, l_leafs_ids)}\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Regions</th>\n",
       "      <th>Excitatory [mm-3]</th>\n",
       "      <th>Inhibitory [mm-3]</th>\n",
       "      <th>Modulatory [mm-3]</th>\n",
       "      <th>Astrocytes [mm-3]</th>\n",
       "      <th>Oligodendrocytes [mm-3]</th>\n",
       "      <th>Microglia [mm-3]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Ectorhinal area/Layer 1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>54217.228792</td>\n",
       "      <td>65.750071</td>\n",
       "      <td>1835.406238</td>\n",
       "      <td>5625.128418</td>\n",
       "      <td>23225.163391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>Ectorhinal area/Layer 2/3</td>\n",
       "      <td>107547.686691</td>\n",
       "      <td>6218.964878</td>\n",
       "      <td>267.907385</td>\n",
       "      <td>3672.527141</td>\n",
       "      <td>18883.517931</td>\n",
       "      <td>42262.609627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>Ectorhinal area/Layer 5</td>\n",
       "      <td>122254.027187</td>\n",
       "      <td>3989.699982</td>\n",
       "      <td>647.703455</td>\n",
       "      <td>6073.657105</td>\n",
       "      <td>36820.408408</td>\n",
       "      <td>36673.812803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Ectorhinal area/Layer 6a</td>\n",
       "      <td>134478.115867</td>\n",
       "      <td>2997.008767</td>\n",
       "      <td>712.944817</td>\n",
       "      <td>5866.391611</td>\n",
       "      <td>54271.823964</td>\n",
       "      <td>29816.056387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Ectorhinal area/Layer 6b</td>\n",
       "      <td>92443.399000</td>\n",
       "      <td>6341.664216</td>\n",
       "      <td>790.355778</td>\n",
       "      <td>2672.155248</td>\n",
       "      <td>68243.457806</td>\n",
       "      <td>18168.773890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949</th>\n",
       "      <td>mammillotegmental tract</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13082.290981</td>\n",
       "      <td>95431.204740</td>\n",
       "      <td>3591.836735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>mammillary peduncle</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21523.131673</td>\n",
       "      <td>93115.065243</td>\n",
       "      <td>683.274021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>stria medullaris</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26794.563438</td>\n",
       "      <td>80137.764179</td>\n",
       "      <td>48436.063474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>fasciculus retroflexus</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9203.914308</td>\n",
       "      <td>63826.500926</td>\n",
       "      <td>36811.425549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>habenular commissure</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>42094.545455</td>\n",
       "      <td>82967.272727</td>\n",
       "      <td>45498.181818</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Regions  Excitatory [mm-3]  Inhibitory [mm-3]  \\\n",
       "270    Ectorhinal area/Layer 1           0.000000       54217.228792   \n",
       "271  Ectorhinal area/Layer 2/3      107547.686691        6218.964878   \n",
       "272    Ectorhinal area/Layer 5      122254.027187        3989.699982   \n",
       "273   Ectorhinal area/Layer 6a      134478.115867        2997.008767   \n",
       "274   Ectorhinal area/Layer 6b       92443.399000        6341.664216   \n",
       "..                         ...                ...                ...   \n",
       "949    mammillotegmental tract           0.000000           0.000000   \n",
       "950        mammillary peduncle           0.000000           0.000000   \n",
       "952           stria medullaris           0.000000           0.000000   \n",
       "953     fasciculus retroflexus           0.000000           0.000000   \n",
       "954       habenular commissure           0.000000           0.000000   \n",
       "\n",
       "     Modulatory [mm-3]  Astrocytes [mm-3]  Oligodendrocytes [mm-3]  \\\n",
       "270          65.750071        1835.406238              5625.128418   \n",
       "271         267.907385        3672.527141             18883.517931   \n",
       "272         647.703455        6073.657105             36820.408408   \n",
       "273         712.944817        5866.391611             54271.823964   \n",
       "274         790.355778        2672.155248             68243.457806   \n",
       "..                 ...                ...                      ...   \n",
       "949           0.000000       13082.290981             95431.204740   \n",
       "950           0.000000       21523.131673             93115.065243   \n",
       "952           0.000000       26794.563438             80137.764179   \n",
       "953           0.000000        9203.914308             63826.500926   \n",
       "954           0.000000       42094.545455             82967.272727   \n",
       "\n",
       "     Microglia [mm-3]  \n",
       "270      23225.163391  \n",
       "271      42262.609627  \n",
       "272      36673.812803  \n",
       "273      29816.056387  \n",
       "274      18168.773890  \n",
       "..                ...  \n",
       "949       3591.836735  \n",
       "950        683.274021  \n",
       "952      48436.063474  \n",
       "953      36811.425549  \n",
       "954      45498.181818  \n",
       "\n",
       "[252 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only the regions that are leafs in the hierarchy to avoid overlap\n",
    "table_cells = table_cells[table_cells['Regions'].isin(l_leafs_names)] \n",
    "table_cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get a dictionnary of lipid expression for each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def compute_array_exp_lipids(l_id_regions, brain_1 = False, decrease_resolution_factor = 5, ):\n",
    "    ll_exp_lipids = []\n",
    "    l_name_lipids = []\n",
    "    # Simulate a click on all lipid names\n",
    "    for name in sorted(\n",
    "        figures._data.get_annotations_MAIA_transformed_lipids(brain_1=brain_1).name.unique()\n",
    "    ):\n",
    "        structures = figures._data.get_annotations_MAIA_transformed_lipids(brain_1=brain_1)[\n",
    "            figures._data.get_annotations_MAIA_transformed_lipids(brain_1=brain_1)[\"name\"] == name\n",
    "        ].structure.unique()\n",
    "        for structure in sorted(structures):\n",
    "            cations = figures._data.get_annotations_MAIA_transformed_lipids(brain_1=brain_1)[\n",
    "                (\n",
    "                    figures._data.get_annotations_MAIA_transformed_lipids(brain_1=brain_1)[\"name\"]\n",
    "                    == name\n",
    "                )\n",
    "                & (\n",
    "                    figures._data.get_annotations_MAIA_transformed_lipids(brain_1=brain_1)[\n",
    "                        \"structure\"\n",
    "                    ]\n",
    "                    == structure\n",
    "                )\n",
    "            ].cation.unique()\n",
    "            for cation in sorted(cations):\n",
    "                l_selected_lipids = []\n",
    "                l_slices = figures._data.get_slice_list(indices=\"brain_1\" if brain_1 else \"brain_2\")\n",
    "                for slice_index in l_slices:\n",
    "\n",
    "                    # Find lipid location\n",
    "                    l_lipid_loc = (\n",
    "                        figures._data.get_annotations()\n",
    "                        .index[\n",
    "                            (figures._data.get_annotations()[\"name\"] == name)\n",
    "                            & (figures._data.get_annotations()[\"structure\"] == structure)\n",
    "                            & (figures._data.get_annotations()[\"slice\"] == slice_index)\n",
    "                            & (figures._data.get_annotations()[\"cation\"] == cation)\n",
    "                        ]\n",
    "                        .tolist()\n",
    "                    )\n",
    "\n",
    "                    # If several lipids correspond to the selection, we have a problem...\n",
    "                    if len(l_lipid_loc) > 1:\n",
    "                        logging.warning(\"More than one lipid corresponds to the selection\")\n",
    "                        l_lipid_loc = [l_lipid_loc[-1]]\n",
    "                    # If no lipid correspond to the selection, set to -1\n",
    "                    if len(l_lipid_loc) == 0:\n",
    "                        l_lipid_loc = [-1]\n",
    "\n",
    "                    # add lipid index for each slice\n",
    "                    l_selected_lipids.append(l_lipid_loc[0])\n",
    "\n",
    "                # Get final lipid name\n",
    "                lipid_string = name + \" \" + structure + \" \" + cation\n",
    "\n",
    "                # If lipid is present in at least one slice\n",
    "                if np.sum(l_selected_lipids) > -len(l_slices):\n",
    "\n",
    "                    # Build the list of mz boundaries for each peak and each index\n",
    "                    lll_lipid_bounds = [\n",
    "                        [\n",
    "                            [\n",
    "                                (\n",
    "                                    float(figures._data.get_annotations().iloc[index][\"min\"]),\n",
    "                                    float(figures._data.get_annotations().iloc[index][\"max\"]),\n",
    "                                )\n",
    "                            ]\n",
    "                            if index != -1\n",
    "                            else None\n",
    "                            for index in [lipid_1_index, -1, -1]\n",
    "                        ]\n",
    "                        for lipid_1_index in l_selected_lipids\n",
    "                    ]\n",
    "                    print(\"getting data for lipid \", lipid_string)\n",
    "                    l_name_lipids.append(lipid_string)\n",
    "                    l_expr = []\n",
    "                    for id_region in l_id_regions:\n",
    "                        #try:\n",
    "                        interpolated_array = figures.compute_3D_volume_figure(ll_t_bounds = lll_lipid_bounds, name_lipid_1=lipid_string,decrease_dimensionality_factor=decrease_resolution_factor,return_interpolated_array=True, structure_guided_interpolation=False, set_id_regions = set([id_region]))\n",
    "                        l_expr.append(np.mean(interpolated_array))\n",
    "                        #except:\n",
    "                        #    print('An error has occured')\n",
    "                        #    l_expr.append(np.nan)\n",
    "                    ll_exp_lipids.append(l_expr)\n",
    "\n",
    "    return np.array(ll_exp_lipids).T, l_name_lipids\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_from_save = False\n",
    "l_name_regions = list(table_cells['Regions'])\n",
    "l_name_cells = table_cells.columns[1:]\n",
    "if not load_from_save:    \n",
    "    array_exp_lipids_brain_2, l_name_lipids_brain_2 = compute_array_exp_lipids(l_id_regions = [dic_names_id[x] for x in table_cells['Regions']], brain_1 = False) \n",
    "else:\n",
    "    array_exp_lipids_brain_2 = np.load('notebooks/cell_type_atlas/data/array_exp_lipids_False.npy')\n",
    "    array_name_lipids_False = np.load('notebooks/cell_type_atlas/data/array_name_lipids_False.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save lipid data in numpy arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save array of lipid expression for brain 2\n",
    "with open('notebooks/cell_type_atlas/data/array_exp_lipids_False.npy', 'wb') as f:\n",
    "    np.save(f, array_exp_lipids_brain_2)\n",
    "\n",
    "# Save corresponding names for brain 2\n",
    "with open('notebooks/cell_type_atlas/data/array_name_lipids_False.npy', 'wb') as f:\n",
    "    np.save(f, np.array(l_name_lipids_brain_2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get cell type data as an array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.00000000e+00, 5.42172288e+04, 6.57500710e+01, 1.83540624e+03,\n",
       "        5.62512842e+03, 2.32251634e+04],\n",
       "       [1.07547687e+05, 6.21896488e+03, 2.67907385e+02, 3.67252714e+03,\n",
       "        1.88835179e+04, 4.22626096e+04],\n",
       "       [1.22254027e+05, 3.98969998e+03, 6.47703455e+02, 6.07365711e+03,\n",
       "        3.68204084e+04, 3.66738128e+04],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 2.67945634e+04,\n",
       "        8.01377642e+04, 4.84360635e+04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 9.20391431e+03,\n",
       "        6.38265009e+04, 3.68114255e+04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 4.20945455e+04,\n",
       "        8.29672727e+04, 4.54981818e+04]])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "array_exp_cells = table_cells.iloc[:,1:].to_numpy()\n",
    "array_exp_cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Make an elastic net regression to explain lipid expression in terms of cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_regression_all_lipids(array_exp_lipids, array_exp_cells):\n",
    "    \"\"\"Compute the elastic net regression coefficients for all lipids.\n",
    "\n",
    "    Returns:\n",
    "        list(list(float)), list(float): List of coefficients (for each lipid) and list of scores\n",
    "            for the elastic net regression explaining lipid expression in terms of gene expression.\n",
    "    \"\"\"\n",
    "    # Define regression as a function for potential parallelization\n",
    "    def compute_regression(index_lipid):\n",
    "        clf = linear_model.ElasticNet(fit_intercept=True, alpha=0.01, positive=False)\n",
    "        clf.fit(array_exp_cells, array_exp_lipids[:, index_lipid])\n",
    "        return [\n",
    "            clf.coef_,\n",
    "            clf.score(array_exp_cells, array_exp_lipids[:, index_lipid]),\n",
    "        ]\n",
    "\n",
    "    # Compute regression for all lipids\n",
    "    l_lipid_indices = list(range(array_exp_lipids.shape[1]))\n",
    "    l_res = [x for x in map(compute_regression, l_lipid_indices)]\n",
    "\n",
    "    # Store the coefficients and the score of the regressions\n",
    "    ll_coef = []\n",
    "    l_score = []\n",
    "    for res in l_res:\n",
    "        ll_coef.append(res[0])\n",
    "        l_score.append(res[1])\n",
    "\n",
    "    # Return result\n",
    "    return np.array(ll_coef), l_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do the LASSO regression\n",
    "array_coef_brain_2, l_score_brain_2 = compute_regression_all_lipids(array_exp_lipids_brain_2, array_exp_cells)\n",
    "array_coef_brain_2_reversed, l_score_brain_2_reversed = compute_regression_all_lipids(array_exp_cells, array_exp_lipids_brain_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save the filtered data from the molecular atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open('notebooks/cell_type_atlas/data/array_coef.npy', 'wb') as f:\n",
    "    np.save(f, array_coef_brain_2)\n",
    "with open('notebooks/cell_type_atlas/data/array_score.npy', 'wb') as f:\n",
    "    np.save(f, np.array(l_score_brain_2))\n",
    "\n",
    "with open('notebooks/cell_type_atlas/data/array_coef_reversed.npy', 'wb') as f:\n",
    "    np.save(f, array_coef_brain_2_reversed)\n",
    "with open('notebooks/cell_type_atlas/data/array_score_reversed.npy', 'wb') as f:\n",
    "    np.save(f, np.array(l_score_brain_2_reversed))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 ('py310')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "vscode": {
   "interpreter": {
    "hash": "4f1f5731aabae6f0842d315b34d4e49ed061c9fdf55bf70e05ea23a61c2a314b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
